server/room_db.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes"
+
+	pb "github.com/googleapis/gapic-showcase/server/genproto"
+	"google.golang.org/genproto/protobuf/field_mask"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func NewRoomDb() RoomDb {
+	return &roomDb{
+		uid:   &uniqID{},
+		token: NewTokenGenerator(),
+
+		mu:    sync.Mutex{},
+		keys:  map[string]int{},
+		rooms: []roomEntry{},
+	}
+}
+
+type RoomDb interface {
+	Create(u *pb.Room) (*pb.Room, error)
+	Get(name string) (*pb.Room, error)
+	Update(u *pb.Room, f *field_mask.FieldMask) (*pb.Room, error)
+	Delete(name string) error
+	List(page_size int32, token string) (*pb.ListRoomsResponse, error)
+}
+
+type ReadOnlyRoomDb interface {
+	Get(name string) (*pb.Room, error)
+}
+
+type roomEntry struct {
+	room    *pb.Room
+	deleted bool
+}
+
+type roomDb struct {
+	uid   *uniqID
```

Comment:
I think this can just be the value, not the pointer, then you can also elide the init in `NewRoomDb`.